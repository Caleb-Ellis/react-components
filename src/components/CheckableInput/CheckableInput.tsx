import classNames from "classnames";
import { TSFixMe } from "index";
import { nanoid } from "nanoid";
import PropTypes from "prop-types";
import React, { useEffect, useRef, HTMLProps } from "react";

type Props = {
  inputType: "radio" | "checkbox";
  label: string;
  required?: boolean;
  indeterminate?: boolean;
} & HTMLProps<HTMLInputElement>;

const CheckableInput = ({
  inputType,
  label,
  required = false,
  indeterminate = false,
  ...checkboxProps
}: Props): JSX.Element => {
  const inputId = useRef(nanoid());
  // setting this type as any for now, otherwise it causes linting errors
  const inputRef: TSFixMe = useRef(null);

  useEffect(() => {
    inputRef.current.indeterminate = indeterminate;
  }, [indeterminate]);

  return (
    <label
      className={classNames(`p-${inputType}`, { "is-required": required })}
    >
      <input
        ref={inputRef}
        type={inputType}
        aria-labelledby={inputId.current}
        className={`p-${inputType}__input`}
        {...checkboxProps}
      />
      <span className={`p-${inputType}__label`} id={inputId.current}>
        {label}
      </span>
    </label>
  );
};

CheckableInput.propTypes = {
  inputType: PropTypes.oneOf(["checkbox", "radio"]).isRequired,
  label: PropTypes.string.isRequired,
  required: PropTypes.bool,
  indeterminate: PropTypes.bool,
};

export default CheckableInput;
