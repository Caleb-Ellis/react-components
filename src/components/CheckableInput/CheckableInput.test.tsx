import { shallow } from "enzyme";
import React from "react";

import CheckableInput from "./CheckableInput";

describe("CheckableInput ", () => {
  it("renders a checkable input", () => {
    const wrapper = shallow(
      <CheckableInput type="checkbox" label="Test label"></CheckableInput>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
