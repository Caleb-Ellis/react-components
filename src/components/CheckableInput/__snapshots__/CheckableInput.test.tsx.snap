// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckableInput  renders a checkable input 1`] = `
<label
  className="p-checkbox"
>
  <input
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-checkbox__input"
    type="checkbox"
  />
  <span
    className="p-checkbox__label"
    id="Uakgb_J5m9g-0JDMbcJqLJ"
  >
    Test label
  </span>
</label>
`;
